File: model_dev_and_vld/bcr.py
import pandas as pd
import numpy as np
import math
from scipy.stats import beta, norm
from scipy.optimize import root_scalar

#data import
fp = "https://andrija-djurovic.github.io/adsfcr/model_dev_and_vld/BCR_TABLES.xlsx"
tbl_1 = pd.read_excel(io = fp, 
                      sheet_name = "Table1")
tbl_2 = pd.read_excel(io = fp, 
                      sheet_name = "Table2")
tbl_3 = pd.read_excel(io = fp, 
                      sheet_name = "Table3")


#optimization function 
def opt_f(pd, n, k, theta, rho, T, cl, N):
    #simulated confidence interval
    cl_sim = 1 - np.mean([ss(pd, n, k, theta, rho, T) for _ in range(N)])
    #simulation error
    dev = cl_sim - cl
    return dev

#single simulation
def ss(pd, n, k, theta, rho, T):
    #systemic factor
    z = np.empty(T)
    z[0] = np.random.normal()
    #correlated systemic factor
    if T > 1:
        for i in range(1, T):
            z[i] = theta * z[i-1] + np.sqrt(1 - theta**2) * np.random.normal()
    #conditional pd
    pdc = (norm.ppf(q = pd) - z * np.sqrt(rho)) / np.sqrt(1 - rho)
    #cumulative probability of default
    pd_c = 1 - np.prod(1 - norm.cdf(x = pdc))
    #likelihood
    lh = binom_cum(n = n, 
                   p = pd_c, 
                   k = k)
    return lh

#likelihood of observing no more than k defaults out of n obligors
'''
#numerical solution
def binom_cum(n, p, k):
    #initialize bc
    bc = 0
    # P(x <= k)
    for i in range(k + 1):
        bc += math.comb(n, i) * (p ** i) * ((1 - p) ** (n - i))
    return bc
'''
#analytical solution
def binom_cum(n, p, k):
    bc = 1 - beta.cdf(p, k + 1, n - k)
    return bc

File: model_dev_and_vld/bcr.R
library(openxlsx)

#data import
fp <- "https://andrija-djurovic.github.io/adsfcr/model_dev_and_vld/BCR_TABLES.xlsx"
tbl.1 <- read.xlsx(xlsxFile = fp, 
                   sheet = "Table1")
tbl.2 <- read.xlsx(xlsxFile = fp, 
                   sheet = "Table2")
tbl.3 <- read.xlsx(xlsxFile = fp, 
                   sheet = "Table3")

#optimization function 
opt.f <- function(pd, n, k, theta, rho, T, cl, N) {
      #simulated confidence interval
      cl.sim = 1 - mean(replicate(n = N, 
                                  ss(pd = pd, 
                                     n = n, 
                                     k = k, 
                                     theta = theta, 
                                     rho = rho, 
                                     T = T)))
      #simulation error
      dev <- cl.sim - cl
return(dev)
}

#single simulation
ss <- function(pd, n, k, theta, rho, T) {
      #systemic factor
      z <- rep(NA, T)
      z[1] <- rnorm(1)
      #correlated systemic factor
      if    (T > 1) {
            for   (j in 2:T) {
                  z[j] <- theta * z[j-1] + sqrt(1 - theta^2) * rnorm(1)
                  }
            }
      #conditional pd
      pdc <- (qnorm(pd) - z*sqrt(rho)) / sqrt(1 - rho)
      #cumulative probability of default
      pd.c <- 1 - prod(1 - pnorm(q = pdc))
      #likelihood 
      lh <- binom.cum(n = n, 
                      p = pd.c, 
                      k = k)
return(lh)
}

#likelihood of observing no more than k defaults out of n obligors
#numerical solution
#binom.cum <-  function (n, p, k) {
#      #initiate bc
#      bc <- 0
#      #P(x <= k)
#      for   (j in 0:k) {
#            bc <-  bc + choose(n = n, k = j) * (p^j) * ((1 - p)^(n - j))
#	      }
#return(bc)
#}
#analytical solution
binom.cum <-  function (n, p, k) {
      bc <- 1 - pbeta(q = p, 
                      shape1 = k + 1, 
                      shape2 = n - k)
return(bc)
}
